2024/BIN-2024-0003.md

```diff
@@ -0,0 +1,124 @@
+| BIN-2024-0003 | `CHECKSIGFROMSTACK`
+| :------------ | :-------
+| Revision      | 000 (2024-01-17)
+| Author        | Brandon Black `<freedom@reardencode.com>`
+| |
+| Layer         | Consensus (soft fork)
+| Status        | Draft
+| License       | BSD-3-CLAUSE
+| |
+| Discussion    | https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376/9
+| Aliases       | [BIPs PR#1535](https://github.com/bitcoin/bips/pull/1535)
+
+## Abstract
+
+This BIP describes two new opcode for the purpose of checking cryptographic
+signatures in bitcoin scripts against data other than bitcoin transactions.
+
+## Summary
+
+We propose replacing `OP_NOP5` (0xb4) in bitcoin script with
+`OP_CHECKSIGFROMSTACKVERIFY`. When verifying taproot script spends having
+leaf version 0xc0 (as defined in BIP342), we propose `OP_CHECKSIGFROMSTACK`
+to replace `OP_SUCCESS204` (0xcc).
+
+`OP_CHECKSIGFROMSTACK` and `OP_CHECKSIGFROMSTACKVERIFY`
+have semantics similar to `OP_CHECKSIG` and
+`OP_CHECKSIGVERIFY` respectively, as specified below.
+
+For legacy and segwit v0 scripts, 32-byte and 33-byte keys are supported for
+BIP340 signature verification and ECDSA signature verification respectively.
+
+## Specification
+
+* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
+* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
+* For `OP_CHECKSIGFROMSTACK` the top three elements are popped from the stack.
+* If the public key size is zero, the script MUST fail and terminate immediately.
+* If the public key size is 32 bytes, it is considered to be a public key as described in BIP340:
+    * If the signature is not the empty vector, the signature is validated against the public key and message according to BIP340. Validation failure in this case immediately terminates script execution with failure.
+* For legacy and segwit v0 scripts, if the public key size is 33 bytes and its first byte is 0x02 or 0x03, it is considered a compressed public key as described in BIP137:
```

Comment:
Not clear to me that it makes sense to support ECDSA at all. If it does, it's not clear to me why it wouldn't be available via a 33 byte pubkey in tapscript.

Reply:
Interesting. My original thought was to provide only the same type of signing as the other sigops on the same script type (with the same semantics); but pretty quickly it became clear that adopting Tapscript key semantics would provide better future upgrade properties, and let me add BIP340 signature verification to legacy scripts (which has several benefits).

I'm honestly open to both or neither for ECDSA, but was hesitant to consume 33-byte keys with 0x02 and 0x03 in Tapscript CSFS right off the hop, thinking that we might want those specific keys for x-y Schnorr signatures in the future.

Reply:
I think it'd be really surprising to use 02/03 prefix with 33-byte keys in tapscript CHECKSIG, seems like asking for trouble. More generally, I think 33-byte keys for tapscript CHECKSIG probably don't have an purpose other than being a 32-byte BIP340 key with a 1-byte flag that changes how the sighash will be interpreted. If that's true, I think there's no conflict with CSFS -- it doesn't make use of a sighash anyway. YMMV though.